src/manager/evm.rs

```diff
@@ -159,20 +249,124 @@ impl EthSubnetManager<MiddlewareImpl> {
-            let provider = Http::new_with_auth(
-                url,
-                Authorization::Bearer(auth_token.unwrap_or_default().to_string()),
-            )?;
+            let provider = if auth_token.is_some() {
+                Http::new_with_auth(
+                    url,
+                    Authorization::Bearer(auth_token.unwrap_or_default().to_string()),
+                )?
+            } else {
+                Http::new(url)
+            };
+
             let provider = Provider::new(provider);
             let wallet = priv_key.parse::<LocalWallet>()?;
             let wallet = wallet.with_chain_id(subnet.chain_id.unwrap_or_default());
 
+            let evm_gateway_address = subnet
+                .evm_gateway_address
+                .as_ref()
+                .ok_or_else(|| anyhow!("evm gateway address not defined"))?;
+            let gateway_address = H160::from_str(evm_gateway_address)?;
+
+            let evm_registry_address = subnet
+                .evm_registry_address
+                .as_ref()
+                .ok_or_else(|| anyhow!("evm registry address not defined"))?;
+            let evm_registry_address = H160::from_str(evm_registry_address)?;
+
             let signer = Arc::new(SignerMiddleware::new(provider, wallet));
-            return Ok(Self::new(signer));
+            let gateway_contract = Gateway::new(gateway_address, Arc::new(signer.clone()));
+            let evm_registry_contract =
+                IPCRegistry::new(evm_registry_address, Arc::new(signer.clone()));
+            return Ok(Self::new(signer, gateway_contract, evm_registry_contract));
         }
 
         Err(anyhow!(
             "no ethereum-compatible private key provided in config"
         ))
     }
 }
+
+// fn evm_id_to_address(evm_subnet: ipc_registry::SubnetID) -> Result<Address> {
+//     // TODO: maybe do a check to ensure the parent subnet id have common parents with new child id
+//     let mut children = vec![];
+//     let addr = evm_subnet
+//         .route
+//         .last()
+//         .ok_or_else(anyhow!("invdalid evm address passed"))?;
+//     let eth_addr = EthAddress::from_str(addr.to_string())?;
+//     Ok(Address::from(eth_addr))
+// }
+
+fn agent_subnet_to_evm_address(subnet: &SubnetID) -> Result<ethers::types::Address> {
+    let children = subnet.children();
+    let ipc_addr = children
+        .last()
+        .ok_or_else(|| anyhow!("{subnet:} has no child"))?;
+
+    payload_to_evm_address(ipc_addr.payload())
+}
+
+fn payload_to_evm_address(payload: &Payload) -> Result<ethers::types::Address> {
+    match payload {
+        Payload::Delegated(delegated) => {
+            let slice = delegated.subaddress();
+            Ok(ethers::types::Address::from_slice(&slice[0..20]))
+        }
+        _ => Err(anyhow!("invalid is invalid")),
+    }
+}
+
+fn agent_subnet_to_evm_addresses(subnet: &SubnetID) -> Result<Vec<ethers::types::Address>> {
```

Comment:
Should we make it a proper From implementation? 
```suggestion
fn subnet_to_evm_address(subnet: &SubnetID) -> Result<Vec<ethers::types::Address>> {
```